language/vm/src/compatibility.rs

```diff
@@ -87,8 +87,12 @@ impl Compatibility {
         }
 
         // old module's public functions are a subset of the new module's public functions
-        for function in &old_module.externally_visible_functions {
-            if !new_module.externally_visible_functions.contains(&function) {
+        //
+        // TODO: the current implementation treats `public(friend)` function the same way as`public`
+        // for compatibility checking. We might need to change this in the future to allow
+        // `public(friend)` functions to update if none of its friends actually calls it.
+        for item in &old_module.externally_visible_functions {
+            if !new_module.externally_visible_functions.contains(&item) {
```

Comment:
We should totally rewrite this. I can put up another PR after this lands as you work on other things

Reply:
I have re-written this in https://github.com/diem/diem/pull/7531/commits/d518d19e907ea64e5ca36ae615824f1c03e84186.